src/main/kotlin/org/rust/ide/injected/RsDoctestLanguageInjector.kt

```diff
@@ -0,0 +1,152 @@
+
+// See https://github.com/rust-lang/rust/blob/5182cc1c/src/librustdoc/html/markdown.rs#L646
+private val LANG_SPLIT_REGEX = Regex("[^a-zA-Z0-9-_]+")
+private val RUST_LANG_ALIASES = listOf(
+    "rust",
+    "allow_fail",
+    "should_panic",
+    "no_run",
+    "test_harness",
+//    "compile_fail", // don't highlight code that is expected to contain errors
+    "edition2018",
+    "edition2015"
+)
+
+class RsDoctestLanguageInjector : MultiHostInjector {
+    override fun elementsToInjectIn(): List<Class<out PsiElement>> =
+        listOf(RsDocCommentImpl::class.java)
+
+    override fun getLanguagesToInject(registrar: MultiHostRegistrar, context: PsiElement) {
+        if (context !is PsiLanguageInjectionHost) return
+        val elementType = context.elementType
+        if (context.isValidHost && elementType in RS_DOC_COMMENTS) {
+            val text = context.text
+            val tripleBacktickIndices = text.indicesOf("```").toList()
+            if (tripleBacktickIndices.size < 2) return // no code blocks in the comment
+
+            val rsElement = context.ancestorStrict<RsElement>() ?: return
+            val cargoTarget = rsElement.containingCargoTarget ?: return
+            if (!cargoTarget.isLib) return // only library targets can have doctests
+            val crateName = cargoTarget.normName
+            val infix = RsDocKind.of(elementType).infix
+
+            tripleBacktickIndices.chunked(2).mapNotNull { idx ->
+                // Contains code lines inside backticks including `///` at the start and `\n` at the end.
+                // It doesn't contain the last line with /// ```
+                val lines = run {
+                    val codeBlockStart = idx[0] + 3
+                    val codeBlockEnd = idx.getOrNull(1) ?: return@mapNotNull null
+                    generateSequence(codeBlockStart) { text.indexOf("\n", it) + 1 }
+                        .takeWhile { it != 0 && it < codeBlockEnd }
+                        .zipWithNext()
+                        .iterator()
+                }
+
+                // ```rust, should_panic, edition2018
+                //     ^ this text
+                val lang = lines.nextOrNull()?.let { text.substring(it.first, it.second - 1) } ?: return@mapNotNull null
+                if (!lang.isEmpty()) {
+                    val parts = lang.split(LANG_SPLIT_REGEX).filter { it.isNotBlank() }
+                    if (parts.any { it !in RUST_LANG_ALIASES }) return@mapNotNull null
+                }
+
+                // skip doc comment infix (`///`, `//!` or ` * `)
+                val ranges = lines.asSequence().mapNotNull { (lineStart, lineEnd) ->
+                    val index = text.indexOf(infix, lineStart)
+                    if (index != -1 && index < lineEnd) {
+                        val start = index + infix.length
+                        Triple(start, lineEnd, CharArrayUtil.shiftForward(text, start, lineEnd, " \t"))
```

Comment:
Maybe introduce a new class instead of triples? Something like
```
class Range(val start: Int, val end: Int, val codeStart: Int) {
    fun isCodeNotEmpty(): Boolean = codeStart + 1 < end

    val codeLength: Int = end - codeStart

    fun withIndent(indent: Int): Range? =
        if (indent < end) Range(codeStart, start + indent, end) else null
}
```
And then use
```
val commonIndent = ranges.filter { it.isCodeNotEmpty() }.map { it.codeLength }.min()
val newRanges = if (commonIndent != null) ranges.mapNotNull { it.withIndent(commonIndent) } else ranges
```